File: tools/source/sdks/java-sdk/examples/WorkerRuntimeRoundtripExample.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WorkerRuntimeRoundtripExample {
    public static void main(String[] args) throws Exception {
        var client = new GameAgentEngineClient(
            System.getenv().getOrDefault("GAE_SERVER", "http://127.0.0.1:8080"),
            System.getenv().getOrDefault("GAE_KEY", "dev-key")
        );

        var consumer = System.getenv().getOrDefault("GAE_CONSUMER", "game_client");
        var owner = System.getenv().getOrDefault("GAE_OWNER", "java-sdk-roundtrip");
        var callbackStatus = System.getenv().getOrDefault("GAE_CALLBACK_STATUS", "success");

        var pendingJson = client.listPendingRuntimeTasks(consumer, 1);
        var taskId = findFirstField(pendingJson, "task_id");
        if (taskId == null || taskId.isBlank()) {
            System.out.println("No pending runtime task for consumer=" + consumer + ".");
            return;
        }

        var interfaceName = defaultString(findFirstField(pendingJson, "interface_name"), "unknown_interface");
        System.out.println("Claiming task " + taskId + " (" + interfaceName + ")");

        var claimedJson = client.claimRuntimeTask(taskId, consumer, owner);
        var leaseToken = findFirstField(claimedJson, "lease_token");
        if (leaseToken == null || leaseToken.isBlank()) {
            throw new IllegalStateException("Task " + taskId + " missing lease token after claim.");
        }

        var startedJson = client.startRuntimeTask(taskId, leaseToken);
        System.out.println(startedJson);

        var callbackId = System.getenv().getOrDefault(
            "GAE_CALLBACK_ID",
            defaultString(findFirstField(startedJson, "callback_id"), findFirstField(pendingJson, "callback_id"))
        );
        if (callbackId == null || callbackId.isBlank()) {
            throw new IllegalStateException("Task " + taskId + " missing callback_id.");
        }

        var callbackRequestId = System.getenv().getOrDefault("GAE_CALLBACK_REQUEST_ID", "java-sdk-" + taskId);
        var resultJson = "{" +
            "\"worker\":\"java-sdk-example\"," +
            "\"source\":\"worker_runtime_roundtrip\"," +
            "\"interface_name\":\"" + jsonEscape(interfaceName) + "\"," +
            "\"task_id\":\"" + jsonEscape(taskId) + "\"," +
            "\"consumer\":\"" + jsonEscape(consumer) + "\"" +
        "}";

        var callbackJson = client.actionCallback(callbackId, callbackStatus, resultJson, callbackRequestId);
        System.out.println(callbackJson);
    }

    private static String findFirstField(String json, String fieldName) {
        var pattern = Pattern.compile("\\\"" + Pattern.quote(fieldName) + "\\\"\\s*:\\s*\\\"([^\\\"]+)\\\"");
        Matcher matcher = pattern.matcher(json);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return null;
    }

    private static String defaultString(String value, String fallback) {
        if (value == null || value.isBlank()) {
            return fallback;
        }
        return value;
    }

    private static String jsonEscape(String value) {
        return value
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\r", "\\r")
            .replace("\n", "\\n");
    }
}

File: tools/source/sdks/cs-sdk/examples/WorkerRuntimeRoundtripExample.cs
using GameAgentEngine.SDK;

static string Required(string name, string? fallback = null)
{
    var value = Environment.GetEnvironmentVariable(name) ?? fallback;
    if (string.IsNullOrWhiteSpace(value))
    {
        throw new InvalidOperationException($"Missing required environment variable: {name}");
    }
    return value;
}

var client = new GameAgentEngineClient(
    Environment.GetEnvironmentVariable("GAE_SERVER") ?? "http://127.0.0.1:8080",
    Environment.GetEnvironmentVariable("GAE_KEY") ?? "dev-key");

var consumer = Environment.GetEnvironmentVariable("GAE_CONSUMER") ?? "game_client";
var owner = Environment.GetEnvironmentVariable("GAE_OWNER") ?? "cs-sdk-roundtrip";
var resultStatus = Environment.GetEnvironmentVariable("GAE_CALLBACK_STATUS") ?? "success";

var pending = await client.ListPendingRuntimeTasksAsync(consumer, 1);
var task = pending?.Tasks?.FirstOrDefault();
if (task is null)
{
    Console.WriteLine($"No pending runtime task for consumer={consumer}.");
    return;
}

Console.WriteLine($"Claiming task {task.TaskId} ({task.InterfaceName ?? "unknown_interface"})");
var claimed = await client.ClaimRuntimeTaskAsync(task.TaskId, consumer, owner);
var claimedTask = claimed?.Task;
if (string.IsNullOrWhiteSpace(claimedTask?.LeaseToken))
{
    throw new InvalidOperationException($"Task {task.TaskId} missing lease token after claim.");
}

var started = await client.StartRuntimeTaskAsync(task.TaskId, claimedTask.LeaseToken);
Console.WriteLine($"Started task {task.TaskId} status={started?.Task?.Status}");

var callbackId = Required("GAE_CALLBACK_ID", started?.Task?.CallbackId ?? task.CallbackId);
var callbackRequestId = Environment.GetEnvironmentVariable("GAE_CALLBACK_REQUEST_ID") ?? $"cs-sdk-{task.TaskId}";
var result = new
{
    worker = "cs-sdk-example",
    source = "worker_runtime_roundtrip",
    interface_name = task.InterfaceName,
    task_id = task.TaskId,
    consumer,
};

var callback = await client.ActionCallbackAsync(callbackId, resultStatus, result, callbackRequestId);
Console.WriteLine($"callback_status={callback?.Status} resume_execution_id={callback?.ResumeExecutionId} resumed={(callback?.Resumed is not null)} post_process_applied={(callback?.PostProcess?.Applied ?? false)}");

